Homework 01 Howard Cen
======================================================================

### Q0 **(0 pts)** Intake

Load the data. Smell test it. Fiddle with factor levels and do any other light cleaning you deem necessary. You can set `include = FALSE` for some/all of this code, but DO IT.
```{r include=FALSE}
setwd("C:/Users/Cen Haoning/Documents/STAT540/")
library(lattice)
library(ggplot2)
library(plyr)
library(limma)
library(RColorBrewer)
library(preprocessCore)
library('reshape2')
source("http://www.bioconductor.org/biocLite.R") 
biocLite("limma") 
biocLite("statmod")
```
Clean up "GSE7191-design.txt" (not shown).
```{r include=FALSE}
library(car)
prDes <- read.table("GSE7191-design.txt", sep = "\t", 
                    header = T, row.names = 1)
names(prDes) <- c("Date", "gType", "bRegion", "Sex")
prDes$sidChar <- as.character(rownames(prDes))
rownames(prDes) <- NULL
prDes$gType <- factor(as.character(prDes$gType), levels = 
                        c("Wild_type", "S1P2_KO", "S1P3_KO"))
prDes$gType <- recode(prDes$gType,"'Wild_type'='wt';'S1P3_KO'='SP3KO';
                      'S1P2_KO'='SP2KO'", levels = c('wt', 'SP3KO', 'SP2KO'))
prDes$bRegion <- recode(prDes$bRegion, "'neocortex'='neo'; 'hippocampus'='hip'")
prDes$Sex <- recode(prDes$Sex, "'male'='M'; 'female'='F'")
prDes <- prDes[c("sidChar", "gType", "bRegion", "Sex", "Date")]
prDes <- prDes[with(prDes, order(gType, bRegion, Sex, Date, sidChar)), ]
write.table(prDes, file = "GSE7191_design.tsv", sep = "\t", 
            quote = FALSE, row.names = FALSE)
dput(prDes, file = "GSE7191_design_DPUT.txt")
saveRDS(prDes, file = "GSE7191_design.rds")
```
Clean up "GSE7191-data.txt"(not shown).
```{r include=FALSE}
prDat <- read.table("GSE7191-data.txt",sep = "\t", 
                    header = T, row.names = 1)
D <- match(prDes$sidChar, names(prDat))
cbind(names(prDat)[D], prDes$sidChar)
prDes <- read.table("GSE7191_design.tsv", as.is = 1:2, header = TRUE)
cbind(design = prDes$sidChar, data = names(prDat))
beta <- match(prDes$sidChar, names(prDat))
cbind(names(prDat)[beta], prDes$sidChar)
identical(names(prDat)[beta], prDes$sidChar)
prDat <- prDat[, beta]
write.table(prDat, file = "GSE7191_data.tsv", sep = "\t", quote = FALSE)
```
Read cleaned data and design files:
```{r}
prDat <- read.table("GSE7191_data.tsv")
prDes <- readRDS("GSE7191_design.rds")
```

### Q1 **(2 points)** What are the basic characteristics of the data and meta-data? 

#### Q1a: How many probes? How many samples?
There are `r ncol(prDat)` samples and `r nrow(prDat)`.
#### Q1b: What is the breakdown of samples for Genotype, BrainRegion, Sex, and DateRun?

For starters, cross-tabulate Genotype and BrainRegion. Then do some cross tabulation involving DateRun and, optionally, Sex to investigate if these "nuisance" factors are confounded with the experimental factors. Make sure you've included counts for all 4 individual factors somewhere, even if it's just in the margins of a cross-tabulation. How do you feel about the experimental design? Hint: `table()` and `addmargins()` will help.
```{r}
A <- table(prDes$gType, prDes$bRegion)
B <- table(prDes$gType, prDes$Sex)
C <- table(prDes$bRegion, prDes$Sex)
D <- table(prDes$gType, prDes$bRegion, prDes$Date)
addmargins(A)
addmargins(B)
addmargins(C)
addmargins(D)
addmargins(with(prDes,table(gType, bRegion)))
```
The amount of sample with different factors are not equal, but close. It would be better if there are same amount of sample with each combination of factors.
#### Q1c: Create a plot showing the gene expression data for one probe.

Use position, panels or facets, and color to convey the Genotype, BrainRegion, and Sex of the samples.

One probe is randomly selected, and the gene expression levels are linked to experimental design factors.
```{r fig.width=9, fig.height=6}
set.seed(111)
(theGene <- sample(1:nrow(prDat), 1))
pDat <- data.frame(prDes, gExp = unlist(prDat[theGene, ]))
oDat <- with(pDat, data.frame(sidChar, gExp, Sex, 
            probeset = factor(rep(c("bRegion", "gType"), each = nrow(pDat))), 
            gExp = c(bRegion, gType))) 
stripplot(~ gExp| probeset, oDat, layout = c(nlevels(oDat$probeset), 1), groups = Sex, auto.key = TRUE)
```
#### Q1d: Report the average expression of the selected probe for all possible combinations of Genotype, BrainRegion and Sex.
```{r}
aggregate(gExp ~ bRegion * Sex * gType, pDat, FUN =mean)
```
### Q2 **(4 points)** Examine the sample correlation matrix.

#### Q2a: Depict the sample-to-sample correlations in a heatmap.

At the very least, order the samples by run date; within run date, you could also sort on other factors. What do you see about batch effects, outliers, etc.? Hint: `cor()` and `heatmap()` are helpful.
```{r fig.width=10, fig.height=10}
prDat.ordered <- prDat[order(prDes$Date,prDes$gType,prDes$bRegion,prDes$Sex)]
prDes.ordered <- prDes[order(prDes$Date,prDes$gType,prDes$bRegion,prDes$Sex), ]
hDat <- cor(prDat.ordered)
rownames(hDat) <- with(prDes.ordered, paste(gType,bRegion,Sex,sidChar,sep="_")) 
jGraysFun <- colorRampPalette(brewer.pal(n =9, "Greys"))
heatmap(hDat, Rowv =NA, Colv =NA, scale="none", margins =c(10,10), col = jGraysFun(256))
```

#### Q2b: Identify the outlier sample.

Try to go beyond merely "eyeballing" it, i.e. make a quantitative statement about how it "sticks out" from the other samples.
```{r fig.width=6, fig.height=8}
m <- as.data.frame(colMeans(hDat))
m$sample <- as.character(rownames(m))
rownames(m) <- NULL
names(m) <- c("meanCor", "sample")
stripplot(sample~meanCor, m)
```
From the stripplot, sample GSM172976, is identified as outliers.

#### Q2c: Examine the outlier in the context of its experimental group.

Which group is the outlier in, in terms of Genotype, BrainRegion, and Sex? Scatter plot these samples against each other and comment on the plot. Try to address overplotting! Hint: `splom ()` from `lattice` is helpful.

```{r}
outlier<- 'GSM172976'
Des.no.outlier<- subset(prDes, sidChar!=outlier)
Des.outlier <- subset(prDes, subset=sidChar==outlier)
head(Des.outlier)
```
The only outlier sample is with SP3KO geotype, from hippocampus brain region, and obtained on 01/16/04.
```{r fig.width=8, fig.height=8}
splom(Des.outlier)
```

### Q3 **(4 points)** Normalization, what it can and can't do.

#### Q3a: Make boxplots.

Make a pair of plots: each plot will show gene expression boxplots, one per sample. Identify the outlier sample with a different color, if you can. Comment on these plots, in general and with respect to the outlier.

Here are the 2 plots:

  * The data as provided
  * The data after you apply quantile normalization.

Hint: `normalize.quantiles()` in the Bioconductor package `preprocessCore` is helpful. So are `graphics::boxplot()`, `lattice::bwplot()` and `ggplot2::geom_boxplot()`.

```{r fig.width=6, fig.height=10}
#Make a new tsv data file in which "probe" is a variable:
Data <- read.table("GSE7191-data.txt",header=T)
write.table(Data, file = "GSE7191_data'.tsv", sep = "\t", quote = FALSE)
dput(Data, file = "GSE7191_data'_DPUT.txt")
saveRDS(Data, file = "GSE7191_data'.rds")
Data <- read.table("GSE7191_data'.tsv")
#Draw boxplot:
mDat <- melt(Data,id="probe",variable.name="sample",value.name="gExp")
bwplot(sample~gExp, mDat)
```
Quantile normalize and draw boxplot:
```{r fig.width=6, fig.height=10}
qDat <- as.matrix(prDat)
normalizedDat <- normalize.quantiles(qDat)
rownames(normalizedDat) <- with(Data, paste(probe))
colnames(normalizedDat) <- with(prDes, paste(sidChar))
rDat <- as.data.frame(normalizedDat)
rDat$probe <- as.character(rownames(rDat))
sDat <- melt(rDat,id="probe",variable.name="sample",value.name="gExp_Nor")
bwplot(sample~gExp_Nor, sDat)
```
#### Q3b: Did normalization fix the outlier?

With the quantile-normalized data, re-make the heatmap of the sample correlation matrix and re-compare the outlier to its experimental group. Is everything OK now?

```{r fig.width=10, fig.height=10}
r <- as.data.frame(normalizedDat)
r.ordered <- r[order(prDes$Date,prDes$gType,prDes$bRegion,prDes$Sex)]
h <- cor(r.ordered)
rownames(h) <- with(prDes.ordered, paste(gType,bRegion,Sex,sidChar,sep="_")) 
jGraysFun <- colorRampPalette(brewer.pal(n =9, "Greys"))
heatmap(h, Rowv =NA, Colv =NA, scale="none", margins =c(10,10), col = jGraysFun(256))
```
Quantile normalized data didn't fix outlier. Correlations remained the same. 
#### Q3c: Form a dataset that omits the outlier and quantile normalize it.
```{r}
Data.no.outlier <- subset (Data, select=-GSM172976)
prDat.no.outlier <- subset (prDat, select=-GSM172976)
Dat.no.outlier <- as.matrix(prDat.no.outlier)
normalizedDat.no.outlier <- normalize.quantiles(Dat.no.outlier)
rownames(normalizedDat.no.outlier) <- with(Data.no.outlier, paste(probe))
colnames(normalizedDat.no.outlier) <- with(Des.no.outlier, paste(sidChar))
```

#### Q3d Re-make the heatmap of the sample correlation matrix, now that the worst outlier is gone. Interpret what you see.
```{r}
nn<- as.data.frame(normalizedDat.no.outlier)
nn.ordered <- nn[order(Des.no.outlier$Date,Des.no.outlier$gType,Des.no.outlier$bRegion,Des.no.outlier$Sex)]
Des.ordered <- Des.no.outlier[order(Des.no.outlier$Date,Des.no.outlier$gType,Des.no.outlier$bRegion,Des.no.outlier$Sex), ]
nn.correlation <- cor(nn.ordered)
rownames(nn.correlation) <- with(Des.ordered, paste(gType,bRegion,Sex,sidChar,sep="_")) 
jGraysFun <- colorRampPalette(brewer.pal(n =9, "Greys"))
heatmap(nn.cor, Rowv =NA, Colv =NA, scale="none", margins =c(10,10), col = jGraysFun(256))
```
After the outlier was removed, the colour intensities of correlations have more contrast between each other, which means the pattern and cluster of correlations between samples are clearer.

#### Q3e: Remake the expression boxplots for all samples before moving on to differential expression analysis.

How many samples remain? 
```{r}
Data.no.outlier <- subset(Data, select=-GSM172976)
mDat.no.outlier <- melt(Data.no.outlier,id="probe",variable.name="sample",value.name="gExp")
bwplot(sample~gExp, mDat.no.outlier)
```
There are 49 sample remained.
### Q4 **(5 points)** For each probe, test if it has differential expression across the three genotypes *within the neocortex brain region only*.

You should be using data with the worst outlier removed, after quantile normalization and, for this question, restricted to neocortex. 

#### Q4a: Write out, in an equation or English or, ideally, both, the model you are fitting.

In the context of that model, what statistical test(s) are you conducting? (You can gloss over the neocortex restriction here.)
```{r}
neoDat.normalized <- subset(nn, select = Des.no.outlier$bRegion == "neo")
neoDat <- subset(prDat.no.outlier, select=Des.no.outlier$bRegion=="neo")
neoDes <- subset(Des.no.outlier, Des.no.outlier$bRegion=="neo")
neoDesMat <- model.matrix(~gType, neoDes)
neoFit <- lmFit(neoDat.normalized, neoDesMat)
neoEBayes  <- eBayes(neoFit)
```
The model is liear model.

#### Q4b: Explore your hits.

Display the expression data for the top 50 probes in a heat map. Order the probes by p-value; order the samples by genotype.
```{r}
neoTop <- topTable(neoEBayes, coef = grep('gType', colnames(coef(neoEBayes))), number = 50)
toplist <- grep(paste(neoTop$ID, collapse="|"), row.names(neoDat.normalized))
neoHeat <-as.matrix(neoDat.normalized[toplist,])
ProbeOrder <- order(neoTop$P.Value)
SampleOrder <- order(neoDes$gType)
heatmap(neoHeat, Rowv = SampleOrder, Colv = ProbeOrder, margins =c(10,10), col = jGraysFun(256))
```

#### Q4c: Count your hits.

How many probes have unadjusted p-value < 1e-3? If we took the top 50 probes as our "hits", what is the (estimated) false discovery rate? How many of these hits do we expect to be false discoveries?
```{r}
probelist <- topTable(neoEBayes, coef = grep('gType', colnames(coef(neoEBayes))),number = Inf)
probelist.unadjusted <- subset(probelist, subset = P.Value < 1e-3)
nrow(probelist.unadjusted)
probelist.unadjusted$P.Value <- order(probelist.unadjusted$P.Value)
(probelist.unadjusted$adj.P.Val[50])
(50 * probelist.unadjusted$adj.P.Val[50])
```
96 probes have unadjusted p-value < 1e-3.
False discovery rate is `r (probelist.unadjusted$adj.P.Val[50])`.
`r (50 * probelist.unadjusted$adj.P.Val[50])` of the hits are expected to be false discoveries.
#### Q4d: Plot the gene expression data for a few top hits and a few boring probes.

What are the p-values associated with these probes? Comment on the plots.


#### Q4e: Find probes where the expression in the S1P3 knockout is different from that of wild type.

How many hits do you find if you control FDR at 0.10?
```{r}
hitS1P3<- topTable(neoEBayes, coef = "gTypeSP3KO", p.value = 0.10, number = Inf)
nrow(hitS1P3)
```
There are 61 hits if FDR is at 0.10.

### Q5 **(5 points)** Differential expression analysis for Genotype * BrainRegion.

You should be using data with the worst outlier removed, after quantile normalization and for both brain regions.

#### Q5a: Fit a 3x2 full factorial model.

Test for any effect of Genotype and/or BrainRegion, i.e. test your model against a model with just an intercept. How many probes have a BH-adjusted p-value, a.k.a. q-value, less than 1e-3?
```{r}
gbMat <- model.matrix(~ gType * bRegion, Des.no.outlier)
gbFit <- lmFit(nn, gbMat)
gbEBayes <- eBayes(gbFit)
gbhit <- topTable(gbEBayes, coef = colnames(coef(gbEBayes))[2:6], p.value = 1e-3, number = Inf)
nrow(gbhit)
```
1524 probes have a BH-adjusted p-valueless than 1e-3.

#### Q5b: Test the null hypothesis that BrainRegion doesn't matter, i.e. that all terms involving BrainRegion are zero.

How many probes have a BH-adjusted p-value less than 0.1?
```{r}
bRegionHits <- topTable(gbEBayes, coef = grep('bRegion', colnames(coef(gbEBayes))), p.value = 0.10, number = Inf)
nrow(bRegionHits)
```
3227 probes have a BH-adjusted p-value less than 0.1.
#### Q5c: Highlight some probes where BrainRegion does and does not matter.

Using the results from Q5b, plot and describe some results for a couple of hits and non-hits.


#### Q5d: Test the null hypothesis that Genotype doesn't matter, i.e. that all terms involving Genotype are zero.

How many probes have a BH-adjusted p-value less than 0.1? Compare these results to those obtained for BrainRegion in Q5b, either based on this count or based on all the p-values. What do you conclude about the relative magnitude of the influence of brain region vs. genotype on gene expression?
```{r}
gTypeHits <- topTable(gbEBayes, coef = grep("gType", colnames(coef(gbEBayes))), p.value = 0.10, number = Inf)
nrow(gTypeHits)
```
141 probes have a BH-adjusted p-value less than 0.1. 
Brain region have bigger influence than genotype on gene expression.
